[armv7hl] error: cast from 'char*' to 'TIMESTAMP_STRUCT*'

Hello,

the build of SOCI v4.0.0 on Fedora for the ARMv7HL architecture has been consistently failing. An example of such a build is https://koji.fedoraproject.org/koji/taskinfo?taskID=45383684

The build log is https://kojipkgs.fedoraproject.org//work/tasks/3727/45383727/build.log
and the error is happening around the middle of that page:
```bash
make[2]: Entering directory '/builddir/build/BUILD/soci-4.0.0/tmpbuild'
[ 16%] Building CXX object src/backends/odbc/CMakeFiles/soci_odbc_static.dir/standard-into-type.cpp.o
cd /builddir/build/BUILD/soci-4.0.0/tmpbuild/src/backends/odbc && /usr/bin/g++  -DBOOST_ALL_NO_LIB -DHAVE_DL=1 -DSOCI_ABI_VERSION=\"4.0\" -I/builddir/build/BUILD/soci-4.0.0/tmpbuild -I/builddir/build/BUILD/soci-4.0.0/include -I/builddir/build/BUILD/soci-4.0.0/tmpbuild/include -I/builddir/build/BUILD/soci-4.0.0/include/private  -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -march=armv7-a -mfpu=vfpv3-d16 -mtune=generic-armv7-a -mabi=aapcs-linux -mfloat-abi=hard -pedantic -Werror -Wno-error=parentheses -Wall -Wextra -Wpointer-arith -Wcast-align -Wcast-qual -Wfloat-equal -Woverloaded-virtual -Wredundant-decls -Wno-long-long -std=gnu++98  -Wno-variadic-macros   -o CMakeFiles/soci_odbc_static.dir/standard-into-type.cpp.o -c /builddir/build/BUILD/soci-4.0.0/src/backends/odbc/standard-into-type.cpp
/builddir/build/BUILD/soci-4.0.0/src/backends/odbc/standard-into-type.cpp: In member function 'virtual void soci::odbc_standard_into_type_backend::post_fetch(bool, bool, soci::indicator*)':
/builddir/build/BUILD/soci-4.0.0/src/backends/odbc/standard-into-type.cpp:176:37: error: cast from 'char*' to 'TIMESTAMP_STRUCT*' {aka 'tagTIMESTAMP_STRUCT*'} increases required alignment of target type [-Werror=cast-align]
  176 |             TIMESTAMP_STRUCT * ts = reinterpret_cast<TIMESTAMP_STRUCT*>(buf_);
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [src/backends/odbc/CMakeFiles/soci_odbc_static.dir/build.make:138:
src/backends/odbc/CMakeFiles/soci_odbc_static.dir/standard-into-type.cpp.o] Error 1
make[2]: Leaving directory '/builddir/build/BUILD/soci-4.0.0/tmpbuild'
make[1]: *** [CMakeFiles/Makefile2:677:
src/backends/odbc/CMakeFiles/soci_odbc_static.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

It seems to be somewhat related to issue #777 

Thanks for the great work on SOCI!

Kind regards

Denis

--- a/src/backends/odbc/standard-into-type.cpp
+++ b/src/backends/odbc/standard-into-type.cpp
@@ -8,6 +8,7 @@
 #define SOCI_ODBC_SOURCE
 #include "soci/soci-platform.h"
 #include "soci/odbc/soci-odbc.h"
+#include "soci-compiler.h"
 #include "soci-cstrtoi.h"
 #include "soci-exchange-cast.h"
 #include "soci-mktime.h"
@@ -173,7 +174,14 @@ void odbc_standard_into_type_backend::post_fetch(
         {
             std::tm& t = exchange_type_cast<x_stdtm>(data_);
 
+            // Our pointer should, in fact, be sufficiently aligned, as we
+            // allocate it on the heap, but gcc on ARMv7hl warns about this not
+            // being the case, so just suppress this warning explicitly here.
+            GCC_WARNING_SUPPRESS(cast-align)
+
             TIMESTAMP_STRUCT * ts = reinterpret_cast<TIMESTAMP_STRUCT*>(buf_);
+
+            GCC_WARNING_RESTORE(cast-align)
 
             details::mktime_from_ymdhms(t,
                                         ts->year, ts->month, ts->day,

